File: Python/FactorialWithMemo.py
# creating the memo variable
memo = {}

def factorial (n):
  if n == 0:
    return 1
  elif n in memo:
    
    # checking if result in memo 
    return memo[n]
  else:
    result = n * factorial(n-1)
    
    # updating the memo 
    memo[n] = result
    return result 

# checking below 
print (factorial(2))
print (factorial(5))

print (memo)

File: JS/FactorialWithMemo.js
// creating the memo variable
var memo = {};

var factorial = function(n) {
    if (n === 0) {
        return 1;
    } else if (memo.hasOwnProperty(n)) {

      // checking if i have this in memo
        return memo[n];
    } else {
        var result = factorial(n - 1) * n;

      // updating the memo
        memo[n] = result;
        return result;
    }
};

factorial(2);
factorial(5);
